### PR TITLE
Set pthread signal mask before pthread creating and reset it after pt…

### PR DESCRIPTION
**Pull request by paul-guo-**
…hread create is done.

Previous code could lead to race condition in theory and is not friendly for
error case handling.

### CODE REVIEW

File: src/backend/cdb/motion/ic_udpifc.c [hunk 1]
```diff
@@ -1337,6 +1337,53 @@ initMutex(pthread_mutex_t *mutex)
...
+	sigaddset(&sigs, SIGHUP);
+	sigaddset(&sigs, SIGINT);
+	sigaddset(&sigs, SIGTERM);
+	sigaddset(&sigs, SIGALRM);
+	sigaddset(&sigs, SIGUSR1);
+	sigaddset(&sigs, SIGUSR2);
+
+	err = pthread_sigmask(SIG_BLOCK, &sigs, old_sigs);
+	if (err != 0)
+		elog(ERROR, "Failed to get pthread signal masks with return value: %d", err);
```
Comment (liming01) [inline src/backend/cdb/motion/ic_udpifc.c]: Because here call elog(), it is better to add assert that it is in the main thread.

Comment (paul-guo-) [reply src/backend/cdb/motion/ic_udpifc.c]: @liming01  elog() already does this.

```
#ifdef USE_ASSERT_CHECKING
    if (IsUnderPostmaster && mainthread() != 0 && !pthread_equal(main_tid, pthread_self()))
    {
#if defined(__darwin__)
        write_log("elog called from thread (OS-X pthread_sigmask is broken: MPP-4923)\n");
#else
        write_log("elog called from thread (%s:%d)\n", filename, lineno);
#endif
    }
#endif
```

**Comment by paul-guo-** [comment]
Thanks for reviewing.

File: src/backend/cdb/motion/ic_udpifc.c (hunk 1 above)
Comment (liming01) [reply src/backend/cdb/motion/ic_udpifc.c]: Got it. Thanks.

**Pull request merged**